app/models/body_treatment.rb

```diff
@@ -0,0 +1,70 @@
+class BodyTreatment < ApplicationRecord
+  has_many :durations, dependent: :destroy
+  has_many :locations, dependent: :destroy
+  has_many :studios, through: :locations
+
+  validates :name, presence: true
+  validates :category, presence: true
+  validates :description, presence: true
+
+  # rubocop:disable Metrics/BlockLength
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Layout/LineLength
+  def self.body_treatments
+    treatments_query = BodyTreatment.select(:name, :category, :description, :city, :address, :duration, :price)
+      .joins(:studios, :durations).all.order(:name, :city, :address, :duration)
+    services = {}
+
+    treatments_query.each do |service|
+      if services[service['name']]
+        if services[service['name']]['city'][service['city']]
+          if services[service['name']]['city'][service['city']]['address'][service['address']]
+            services[service['name']]['city'][service['city']]['address'][service['address']]['duration'][service['duration']] =
+              { 'price' => service['price'] }
+          else
+            duration_hash = { service['duration'] => { 'price' => service['price'] } }
+            services[service['name']]['city'][service['city']]['address'][service['address']] =
+              { 'duration' => duration_hash }
+          end
+        else
+          address_hash = {}
+          duration_hash = {}
+
+          duration_hash[service['duration']] = { 'price' => service['price'] }
+          address_hash[service['address']] = { 'duration' => duration_hash }
+          city_hash = { 'address' => address_hash }
+
+          services[service['name']]['city'][service['city']] = city_hash
+        end
+      else
+        service_hash = {}
+        city_hash = {}
+        address_hash = {}
+        duration_hash = {}
+
+        service_hash['category'] = service['category']
+        service_hash['description'] = service['description']
+
+        duration_hash[service['duration']] = { 'price' => service['price'] }
+        address_hash[service['address']] = { 'duration' => duration_hash }
+        city_hash[service['city']] = { 'address' => address_hash }
+        service_hash['city'] = city_hash
+
+        services[service['name']] = service_hash
+      end
+    end
+    services_array = []
+
+    services.each do |key, value|
+      value['name'] = key
+      services_array.push(value)
+    end
+
+    services_array
+  end
```

Comment:
I still don't understand the need for this in the api. The api should ideally be independent of the frontend. All this formatting of how some attributes are grouped under certain keys should be done by the frontend (if even). It also does not bode well that some rubo-cop rules have to be disabled to make this work.

Reply:
Honestly, I'm surprised and I'm tired already explaining it. Israel sends me review reports, which are almost unreadable and now it comes out, he even isn't the reviewer. I'll try one last time more. If I'm just returning the query, it's an array of 123 database lines, instead of an array of 9 separate service objects, which have all the choosing options included as sub hashes. Is the booking service form options selection boxes not understandable, have you tried to book a treatment? I hope, the images are viewable.

![unnamed](https://user-images.githubusercontent.com/22595223/83984486-1b576400-a93e-11ea-9c6a-68a7e9fbfb41.png)
![unnamed (1)](https://user-images.githubusercontent.com/22595223/83984507-31652480-a93e-11ea-8a69-e2ada18f66cc.png)